### PR TITLE
Fix/dataset index: Index values were faulty when indexing the original samples instead of blocks.

### PR DESCRIPTION
**Pull request by le1nux**
# What does this PR do?
The index values in the pbin files had the wrong values. They did start with an offset and additionally, we added another offset of HEADER size when reading from the file buffer. 
See here for the initial offset during pbin index creation:
https://github.com/Modalities/modalities/blob/4aa2e88efe13c3eaab4c6b425fdb82caf0d2a443/src/modalities/dataloader/create_packed_data.py#L145

and the additional offset that is used when reading from the memmap during training:

https://github.com/Modalities/modalities/blob/4aa2e88efe13c3eaab4c6b425fdb82caf0d2a443/src/modalities/dataloader/create_packed_data.py#L262

This PR fixes this issue and makes the index always start at byte 0, only applying the offset once when reading from the memmap file.

## General changes
* index tuples are now always in bytes and the start of the first sample in the data section starts at byte 0 (before the was a wrong offset)
* added test for indexing the original samples 
* the documentation was updated accordingly
* removed `block_size` from abstract classes that don't need to see the `block_size` concept

## Breaking Changes
* None

## Checklist before submitting final PR
- [x] My PR is minimal and addresses one issue / enhancement in isolation
- [x] I have merged main into this feature branch
- [x] I have reviewed my own code w.r.t. correct implementation, missing type hints, proper documentation, etc.
- [x] I have run a sample config for model training
- [x] I have fixed all failing tests (`python tests/tests.py`)

### CODE REVIEW

**Comment by le1nux** [comment]
fixes #163

File: src/modalities/dataloader/create_packed_data.py [hunk 1]
```diff
@@ -161,7 +161,10 @@ def _write_batch(
                         EmbeddedStreamData.TOKEN_SIZE_DESCRIPTOR_LENGTH_IN_BYTES, byteorder="little"
                     )
                 )
-                curr_offset = EmbeddedStreamData.HEADER_SIZE_IN_BYTES
+                # The offset only applies to the data section, not the header
+                # When we load the file, we addtionally add the header size
```
Comment (flxst) [inline src/modalities/dataloader/create_packed_data.py:165]:
```suggestion
                # When we load the file, we additionally add the header size
```

**Review by flxst — APPROVED** [review]
Nice work! I have two general remarks:

1. I think that the class inheritance structure, `PackedMemMapDatasetBase` -> `PackedMemMapDatasetBase`, `PackedMemMapDatasetContinuous` is suboptimal now that the parent class `PackedMemMapDatasetBase` is not an abstract base class anymore, but actually can be instantiated and used by itself. It might be cleaner to restore an abstract base class with an abstract method `_generate_packing_index`, and inherit three classes from it with the three different implementations of the method.
2. The names of the classes are hard to understand and maybe a bit misleading. For instance, the class `PackedMemMapDatasetBase` does not actually do packing, does it? Also, why `Continuous`? Maybe we could try to find better names.

I will approve the PR now since I don't believe that my above suggestions for improvement are essential, but I am happy to take another look if you decide to implement them!

**Comment by le1nux** [comment]
Yes, the inheritance structure can be improved. I suggest we do this in a separate PR together with improving the "packing" terms in those cases when there is no actual packing happening. 

I added the issue https://github.com/Modalities/modalities/issues/167 for addressing this.

File: src/modalities/dataloader/dataset.py [hunk 2]
```diff
@@ -153,23 +149,38 @@ def __init__(self, raw_data_path: Path, block_size: int, sample_key: str):
         self._index = self._generate_packing_index()
 
     def _generate_packing_index(self) -> List[Tuple[int, int]]:
-        raise NotImplementedError
+        # index is a tuple of offset and length in bytes
+        return self._embedded_stream_data.index_base
```
[the remark this answers is not in the document]
Comment (le1nux) [reply src/modalities/dataloader/dataset.py:153]: will be adressed as part of a new PR and issue https://github.com/Modalities/modalities/issues/167

**Review by mali-git — APPROVED** [review]

**Pull request merged**